packages/bsky/src/services/indexing/index.ts

```diff
@@ -93,6 +103,81 @@ export class IndexingService {
     }
   }
 
+  async indexRepo(did: string, commit: string) {
+    const now = new Date().toISOString()
+    const { pds, signingKey } = await this.didResolver.resolveAtpData(did)
+    const { api } = new AtpAgent({ service: pds })
+
+    const { data: car } = await retryHttp(() =>
+      api.com.atproto.sync.getCheckout({ did, commit }),
+    )
+    const { root, blocks } = await readCarWithRoot(car)
+    const storage = new MemoryBlockstore(blocks)
+    const checkout = await verifyCheckoutWithCids(
+      storage,
+      root,
+      did,
+      signingKey,
+    )
+
+    // First, replace the user with just their current profile if it exists
+    await this.db.transaction(async (tx) => {
+      const indexingTx = new IndexingService(tx, this.didResolver)
+      await indexingTx.deleteForUser(did)
+      const profile = checkout.contents[ids.AppBskyActorProfile]?.self
+      if (profile) {
+        const profileUri = AtUri.make(did, ids.AppBskyActorProfile, 'self')
+        await indexingTx.indexRecord(
+          profileUri,
+          profile.cid,
+          profile.value,
+          WriteOpAction.Create,
+          now,
+        )
+      }
+    })
+
+    // Then iterate over all records and index them in batches
+    const contentList = [...walkContentsWithCids(checkout.contents)]
+    const chunks = chunkArray(contentList, 100)
+
+    for (const chunk of chunks) {
+      await this.db.transaction(async (tx) => {
```

Comment:
sorry to come back to this, but what's the recovery path if this part fails?

the user is then left with no data except their profile, right?

Reply:
a bit worried that a bad commit can blow up a user's account? would be better for it just to not update I think

Reply:
oh wait are these nested txs? if so, mind walking me through the different cases here? I'm not super versed in nested txs

Reply:
Cool cool!

To get more concrete—the issue I'm picturing with processing partial writes within a commit is that you might butt into conflicts w/ application semantics.  Switching a downvote to an upvote, for example, might involve removing a downvote and creating an upvote, and doing half of that process isn't going to land you in a consistent state.

This same issue seems much less immediate with processing a full checkout, which consists of only creates, rather than updates/deletes which are far more likely to pair with other ops that are expected to be applied in tandem.

Reply:
It's kinda tricky tho cause at least in that case, I think I would prefer for my downvote to go away even if my upvote is not created. Similarly if I try to batch a bunch of creates & deletes in a single commit as we did during the app migrations. 

That transactionality def seems important to me at the data layer, but I wonder if at the view layer it's best to "just get as close as possible".

I can think of cases where you may prefer transactionality at the view layer, but I can think of more where you'd prefer lenient handling 

Idk it's a tricky question fer sure

Reply:
I tidied this area up and added some logic to skip over invalid records when doing a full repo sync.